### PR TITLE
[feat] Enabled param p for each augmentation

### PR DESCRIPTION
**Pull request by shijianjian**
### Description

#656 

Change log:
1. Enabled ```p``` param to control each augmentation. Mostly element-wise controlling apart from:
    - ```Random*Crop``` APIs, since no ragged tensor supported in PyTorch currently.
    - ```RandomMix*``` APIs, to maximize the variance of the transformation.
    - ```RandomMotionBlur``` API, low-level APIs do not support batch mode.
1. Added ```p_batch``` param to AugmentationBase to control base-wise augmentation. (Note: **invisible to the normal user**. Only given a choice for the development of the new augmentations).
1. Moved ```batch_prob``` generation to AugmentationBase to control the augmentation internally. Performance improvement by only augment the selected inputs in a batch. (Note: **invisible to the normal user**.)
2. Refactored ```random_*_generator``` to generate param only. Removed ```batch_prob``` and some ```flags``` generation. 
2. Refactored ```random_prob_generator``` to use Bernoulli distribution than Uniform.
4. Refactored ```apply_*``` to accept params and flags.
5. Deprecated functional random augmentations ```random_*```. We would encourage using module-like functions for randomized behaviour.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -1,77 +1,22 @@
...
 from . import random_generator as rg
+from .random_generator import AugParamDict
 from .utils import (
-    _infer_batch_shape,
     _range_bound,
     _singular_range_check
 )
 
 
-class AugmentationBase(nn.Module):
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:17]: can you make an alias of this and launch a deprecation warning ?

File: kornia/augmentation/base.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                                      wont be concatenated.
+        same_on_batch (bool): apply the same transformation across the batch. Default: False
+    """
+
+    def infer_batch_shape(self, input: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]) -> torch.Size:
+        """Return the shape of the input. Accepts tuple or tensor."""
+        return _infer_batch_shape(input)
+
+    def transform_tensor(self, input: torch.Tensor) -> torch.Tensor:
+        """Convert any incoming (D, H, W), (C, D, H, W) and (B, C, D, H, W) into (B, C, D, H, W)."""
```
Comment (edgarriba) [inline kornia/augmentation/base.py:224]: is it true ?

File: kornia/augmentation/random_generator/types.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from typing import Dict, Optional
+
+import torch
+try:
+    from typing import TypedDict # type: ignore  # >=3.8  
+except ImportError:
+    from mypy_extensions import TypedDict  # type: ignore  # <=3.7
+
+
+class AugParamDict(TypedDict, total=False):  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/random_generator/types.py:10]: First, I would rename to `AugmentationParameters` to make it more clear since it's not derived from dict. Second, take look at python `dataclasses`, I think it's supported in python >3.6 , double check if there's any workaround on this. Last, why do you `type: ignore` ? this might hurt jitting which is something that we should start to implement.

File: kornia/augmentation/functional/functional.py [hunk 4]
```diff
@@ -44,13 +51,8 @@ def random_hflip(input: torch.Tensor, p: float = 0.5, return_transform: bool = F
...
     """
-    input = _transform_input(input)
-    batch_size, _, h, w = input.size()
-    params = rg.random_prob_generator(batch_size, p=p)
-    output = apply_hflip(input, params)
-    if return_transform:
-        return output, compute_hflip_transformation(input, params)
-    return output
+    raise NotImplementedError(
+        f"functional random augmentation is deprecated. Please use `RandomHorizontalFlip`")
```
Comment (edgarriba) [inline kornia/augmentation/functional/functional.py:55]: I'm fine simplifying the API, but shouldn't we first add the deprecation warning and later remove ? What is the policy to remove functionals ? because I've seen you don't remove all of them.

**Review by edgarriba — COMMENTED** [review]
quick read and added some comments

File: kornia/augmentation/random_generator/types.py (hunk 3 above)
Comment (shijianjian) [reply kornia/augmentation/random_generator/types.py:10]: Renaming is reasonable. ```TypedDict``` will be compiled as normal ```dict``` at the runtime, so it will not cause any issues there. While the ```type: ignore``` is a versioning issue that ```total=False``` is only supported in 3.8 or so.

The reason that I did not go for another class is that I am not sure if we wish to make user construct params like ```from kornia.augmentation import AugmentationParams; param=AugmentationParams(height=1, width=1)```. Would it be overcomplicated?

File: kornia/augmentation/functional/functional.py (hunk 4 above)
Comment (shijianjian) [reply kornia/augmentation/functional/functional.py:55]: I only removed those ```random_*``` stuff. Throw a deprecation warning is not easy here since they are hard to maintain if we put some logic inside the ```AugmentationBase``` modules.

I am still working on these whole things that I haven't decided regarding the best practice here.

**Comment by edgarriba** [comment]
@shijianjian I just feel that this PR includes to many changes without notification to the user. Can we make this smoothly ? I'm in to simplify stuff since the whole augmentations API is getting more and more complex and we need to take care that everything is well documented and tested as seen in recent bug fixes for the return transforms. I vote to discuss in slack or any other public place before introducing such "big" changes.

**Comment by shijianjian** [comment]
@edgarriba Yes, I am aware of that. I got a simpler solution here and it could keep our previous APIs untouched better. Still working on it.

As for the test, can you fix the doctest PR? I am going to add the missing tests in the next PR or so.

File: kornia/augmentation/augmentation.py [hunk 5]
```diff
@@ -259,111 +211,104 @@ class RandomGrayscale(AugmentationBase):
...
+        super(RandomGrayscale, self).__init__(p=p, return_transform=return_transform, same_on_batch=same_on_batch)
 
     def __repr__(self) -> str:
-        repr = f"(p={self.p}, return_transform={self.return_transform}, same_on_batch={self.same_on_batch})"
-        return self.__class__.__name__ + repr
+        return self.__class__.__name__ + f"({super().__repr__()})"
 
     def generate_parameters(self, batch_shape: torch.Size) -> Dict[str, torch.Tensor]:
-        return rg.random_prob_generator(batch_shape[0], self.p, self.same_on_batch)
+        return dict()
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:221]: is there any way that in case we don't override the parent class returns an empty dict ?

File: kornia/augmentation/base.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    _infer_batch_shape,
+    _infer_batch_shape3d,
+    _transform_input,
+    _transform_input3d,
+    _validate_input_dtype,
+    _adapted_sampling
+)
+
+
+class _BasicAugmentationBase(nn.Module):
```
Comment (edgarriba) [inline kornia/augmentation/base.py:21]: I suggest to add minimal test on this classes here

**Review by edgarriba — APPROVED** [review]
LGTM. FeSome suggestions and possibly add few more tests

File: kornia/augmentation/__init__.py [hunk 7]
```diff
@@ -1,5 +1,9 @@
+from .base import (
+    AugmentationBase2D,
+    AugmentationBase3D,
+    MixAugmentationBase
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/__init__.py:4]: I am pretty neutral regarding this. I could remove it for now.

File: kornia/augmentation/augmentation.py [hunk 8]
```diff
@@ -1,77 +1,42 @@
...
 )
 
 
-class AugmentationBase(nn.Module):
-    r"""AugmentationBase base class for customized augmentation implementations. For any augmentation,
-    the implementation of "generate_parameters" and "apply_transform" are required while the
+class AugmentationBase(AugmentationBase2D):
+    r"""AugmentationBase3D base class for customized augmentation implementations.
+
+    For any augmentation, the implementation of "generate_parameters" and "apply_transform" are required while the
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:19]: Yep. I am using "__doc__ = AugmentationBase2D.__doc__" now.

File: kornia/augmentation/augmentation.py (hunk 5 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:221]: I think having this can be safer with knowing it is implemented but expected to be empty.

**Comment by edgarriba** [comment]
@shijianjian rebase and fix the conflicts introduced in 4d0b4f33e5c14a44e9312ad9f0ca86747a9c45cc. Mainly I fixed sphinx warnings, added bibtex citations and rephrased some of the augmentations docs descriptions. After looking at the docs I feel that some of them are quite long and could be simplified a bit so that the user gets a quick intuition about the algorithm and its usage.

File: kornia/augmentation/augmentation.py [hunk 9]
```diff
@@ -23,34 +23,31 @@ def __init__(self, return_transform: bool = False, same_on_batch: bool = False,
 
 
 class RandomHorizontalFlip(AugmentationBase2D):
-    r"""Horizontally flip a tensor image or a batch of tensor images randomly with a given probability.
 
+    r"""Applies a random horizontal flip to a tensor image or a batch of tensor images randomly with a given probability.
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:27]: remove last randomly word (I forgot it). Check others too.
`Applies a random horizontal flip to a tensor image or a batch of tensor images with a given probability.`
